File: go/scanner/scanner.go
package scanner

import (
	"strconv"

	"github.com/AsazuTaiga/crafting_interpriters/go/logger"
	"github.com/AsazuTaiga/crafting_interpriters/go/token"
)

var keywords = map[string]token.TokenType{
	"and":    token.AND,
	"class":  token.CLASS,
	"else":   token.ELSE,
	"false":  token.FALSE,
	"for":    token.FOR,
	"fun":    token.FUN,
	"if":     token.IF,
	"nil":    token.NIL,
	"or":     token.OR,
	"print":  token.PRINT,
	"return": token.RETURN,
	"super":  token.SUPER,
	"this":   token.THIS,
	"true":   token.TRUE,
	"var":    token.VAR,
	"while":  token.WHILE,
}

type Scanner struct {
	source  string
	tokens  []*token.Token
	start   int
	current int
	line    int
}

func NewScanner(source string) *Scanner {
	return &Scanner{
		source: source,
		tokens: make([]*token.Token, 0),
		start:  0,
		current: 0,
		line: 1,
	}
}

func (s *Scanner) ScanTokens(logger *logger.Logger) []*token.Token {
	for !s.isAtEnd() {
		s.start = s.current
		s.scanToken(logger)
	}

	s.tokens = append(s.tokens, token.NewToken(token.EOF, "", nil, s.line))
	return s.tokens
}

func (s *Scanner) scanToken(logger *logger.Logger) {
	c := s.advance();
	switch c {
		case '(':
				s.addToken(token.LEFT_PAREN);
				break;
		case ')':
			s.addToken(token.RIGHT_PAREN);
			break;
		case '{':
			s.addToken(token.LEFT_BRACE);
			break;
		case '}':
			s.addToken(token.RIGHT_BRACE);
			break;
		case ',':
			s.addToken(token.COMMA);
			break;
		case '.':
			s.addToken(token.DOT);
			break;
		case '-':
			s.addToken(token.MINUS);
			break;
		case '+':
			s.addToken(token.PLUS);
			break;
		case ';':
			s.addToken(token.SEMICOLON);
			break;
		case '*':
			s.addToken(token.STAR);
			break;
		case '!':
			if s.match('=') {
				s.addToken(token.BANG_EQUAL);
			} else {
				s.addToken(token.BANG);
			}
			break;
		case '=':
			if s.match('=') {
				s.addToken(token.EQUAL_EQUAL)
			} else {
				s.addToken(token.EQUAL)
			}
			break;
		case '<':
			if s.match('=') {
				s.addToken(token.LESS_EQUAL)
			} else {
				s.addToken(token.LESS)
			}
			break;
		case '>':
			if s.match('=') {
				s.addToken(token.GREATER_EQUAL)
			} else {
				s.addToken(token.GREATER)
			}
			break;
		case '/':
			if s.match('/') {
				for s.peek() != '\n' && !s.isAtEnd() {
					s.advance()
				}
			} else {
				s.addToken(token.SLASH)
			}
			break;
		case ' ':
		case '\r':
		case '\t':
			break;
		case '\n':
			s.line++
			break;
		case '"':
			s.string(logger)
			break;
		case 'o':
			if s.match('r') {
				s.addToken(token.OR)
			}
			break;
		default:
			if s.isDigit(c) {
				s.number()
			} else if s.isAlpha(c) {
				s.identifier()
			} else {
				logger.ErrorReport(s.line, "Unexpected character.");
			}
			break;
	}
}

func (s *Scanner) identifier() {
	for s.isAlphaNumeric(s.peek()) {
		s.advance()
	}

	text := s.source[s.start:s.current]
	if t, ok := keywords[text]; ok {
		s.addToken(t)
	} else {
		s.addToken(token.IDENTIFIER)
	}
}

func (s *Scanner) number() {
	for s.isDigit(s.peek()) {
		s.advance()
	}

	if  s.peek() == '.' && s.isDigit(s.peekNext()) {
		s.advance()

		for s.isDigit(s.peek()) {
			s.advance()
		}
	}

	val, err := strconv.ParseFloat(s.source[s.start:s.current], 64)
	if err != nil {
		panic(err) // NOTE: panicがいいのかはわかってない
	}
	s.addTokenWithLiteral(token.NUMBER, val)
}

func (s *Scanner) string(logger *logger.Logger) {
	for s.peek() != '"' && !s.isAtEnd() {
		if s.peek() == '\n' {
			s.line++
		}
		s.advance()
	}

	if s.isAtEnd() {
		logger.ErrorReport(s.line, "Unterminated string.")
		return
	}

	s.advance()

	value := s.source[s.start+1:s.current-1]
	s.addTokenWithLiteral(token.STRING, value)
}

func (s *Scanner) match(expected byte) bool {
	if s.isAtEnd() {
		return false
	}
	if s.source[s.current] != expected {
		return false
	}

	s.current++
	return true
}

func (s *Scanner) peek() byte {
	if s.isAtEnd() {
		return '\u0000'
	} else {
		return s.source[s.current]
	}
}

func (s *Scanner) peekNext() byte {
	if s.current + 1 >= len(s.source) {
		return '\u0000'
	} else {
		return s.source[s.current+1]
	}
}

func (s *Scanner) isAlpha(c byte) bool {
	return (c >= 'a' && c <= 'z') ||
		(c >= 'A' && c <= 'Z') ||
		c == '_'
}

func (s *Scanner) isAlphaNumeric(c byte) bool {
	return s.isAlpha(c) || s.isDigit(c)
}

func (s *Scanner) isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

func (s *Scanner) isAtEnd() bool {
	return s.current >= len(s.source)
}

func (s *Scanner) advance() byte {
	s.current++
	return s.source[s.current-1]
}

func (s *Scanner) addToken(t token.TokenType)   {
	s.addTokenWithLiteral(t, nil)
}

func (s *Scanner) addTokenWithLiteral(t token.TokenType, literal interface{}) {
	text := s.source[s.start:s.current]
	s.tokens = append(s.tokens, token.NewToken(t, text, literal, s.line))
}

File: php/lox/Scanner.php
<?php

declare(strict_types=1);

namespace Lox;

class Scanner
{
    /** @var $tokens Token[] */
    private array $tokens = [];
    private int $start = 0;
    private int $current = 0;
    private int $line = 1;
    private static array $keywords = [
        "and" => TokenType::AND,
        "class" => TokenType::CLAZZ,
        "else" => TokenType::ELSE,
        "false" => TokenType::FALSE,
        "for" => TokenType::FOR,
        "fun" => TokenType::FUN,
        "if" => TokenType::IF,
        "nil" => TokenType::NIL,
        "or" => TokenType::OR,
        "print" => TokenType::PRINT,
        "return" => TokenType::RETURN,
        "super" => TokenType::SUPER,
        "this" => TokenType::THIS,
        "true" => TokenType::TRUE,
        "var" => TokenType::VAR,
        "while" => TokenType::WHILE,
    ];
    public function __construct(
        private string $source,
    ){}

    /**
     * @return Token[]
     */
    public function scanTokens(): array {
        while (!$this->isAtEnd()) {
            // We are at the beginning of the next lexeme.
            $this->start = $this->current;
            $this->scanToken();
        }

        $this->tokens[] = new Token(TokenType::EOF, "", null, $this->line);
        return $this->tokens;
    }

    private function scanToken(): void
    {
        $c = $this->advance();
        switch ($c) {
            case '(': $this->addToken(TokenType::LEFT_PAREN);
                break;
            case ')': $this->addToken(TokenType::RIGHT_PAREN);
                break;
            case '{': $this->addToken(TokenType::LEFT_BRACE);
                break;
            case '}': $this->addToken(TokenType::RIGHT_BRACE);
                break;
            case ',': $this->addToken(TokenType::COMMA);
                break;
            case '.': $this->addToken(TokenType::DOT);
                break;
            case '-': $this->addToken(TokenType::MINUS);
                break;
            case '+': $this->addToken(TokenType::PLUS);
                break;
            case ';': $this->addToken(TokenType::SEMICOLON);
                break;
            case '*': $this->addToken(TokenType::STAR);
                break;
            case '!':
                $this->addToken(
                    $this->match('=')
                        ? TokenType::BANG_EQUAL
                        : TokenType::BANG
                );
                break;
            case '=':
                $this->addToken(
                    $this->match('=')
                        ? TokenType::EQUAL_EQUAL
                        : TokenType::EQUAL
                );
                break;
            case '<':
                $this->addToken(
                    $this->match('=')
                        ? TokenType::LESS_EQUAL
                        : TokenType::LESS
                );
                break;
            case '>':
                $this->addToken(
                    $this->match('=')
                        ? TokenType::GREATER_EQUAL
                        : TokenType::GREATER
                );
                break;
            case '/':
                if ($this->match('/')) {
                    // A comment goes until the end of the line.
                    while ($this->peek() != '\n' && !$this->isAtEnd()) $this->advance();
                } else {
                    $this->addToken(TokenType::SLASH);
                }
                break;

            case ' ':
            case '\r':
            case '\t':
                // Ignore whitespace.
                break;

            case '\n':
                $this->line++;
                break;

            case '"': $this->string(); break;
            default:
                if ($this->isDigit($c)) {
                    $this->number();
                } else if ($this->isAlpha($c)) {
                    $this->identifier();
                } else {
                    Lox::error($this->line, 'Unexpected character.');
                }
                break;
        }
    }

    private function identifier(): void
    {
        while ($this->isAlphaNumeric($this->peek())) $this->advance();
        $text = substr($this->source, $this->start, $this->current);
        $type = self::$keywords[$text] ?? TokenType::IDENTIFIER;
        $this->addToken($type);
    }

    private function number(): void
    {
        while ($this->isDigit($this->peek())) $this->advance();

        // Look for a fractional part.
        if ($this->peek() == '.' && $this->isDigit($this->peekNext())) {
            // Consume the "."
            $this->advance();

            while ($this->isDigit($this->peek())) $this->advance();
        }

        $this->addToken(
            TokenType::NUMBER,
            floatval(substr($this->source, $this->start, $this->current))
        );
    }

    private function string(): void
    {
        while ($this->peek() != '"' && !$this->isAtEnd()) {
            if ($this->peek() == '\n') $this->line++;
            $this->advance();
        }

        if ($this->isAtEnd()) {
            Lox::error($this->line, "Unterminated string.");
            return;
        }

        // The closing ".
        $this->advance();

        // Trim the surrounding quotes.
        $value = substr($this->source, $this->start + 1, $this->current - 1);
        $this->addToken(TokenType::STRING, $value);
    }

    private function match(string $expected): bool
    {
        if ($this->isAtEnd()) return false;
        if ($this->source[$this->current] !== $expected) return false;

        $this->current++;
        return true;
    }

    private function peek(): string
    {
        if ($this->isAtEnd()) return '\0';
        return $this->source[$this->current];
    }

    private function peekNext(): string
    {
        if ($this->current + 1 >= $this->source) return '\0';
        return $this->source[$this->current + 1];
    }

    private function isAlpha(string $c): bool
    {
        return ($c >= 'a' && $c <= 'z') ||
            ($c >= 'A' && $c <= 'Z') ||
            $c == '_';
    }

    private function isAlphaNumeric(string $c): bool
    {
        return $this->isAlpha($c) || $this->isDigit($c);
    }

    private function isDigit(string $c): bool
    {
        return $c >= '0' && $c <= '9';
    }

    private function isAtEnd(): bool
    {
        return $this->current >= strlen($this->source);
    }

    private function advance(): string
    {
        return $this->source[$this->current++];
    }

    private function addToken(TokenType $type, $literal = null): void
    {
        $text = substr($this->source, $this->start, $this->current);
        $this->tokens[] = new Token($type, $text, $literal, $this->line);
    }
}
